rust/connlib/libs/tunnel/src/tun_darwin.rs

```diff
@@ -0,0 +1,279 @@
+    sockaddr_ctl, sockaddr_in, socket, socklen_t, AF_INET, AF_INET6, AF_SYSTEM, AF_SYS_CONTROL,
+    CTLIOCGINFO, F_GETFL, F_SETFL, IF_NAMESIZE, IPPROTO_IP, O_NONBLOCK, PF_SYSTEM, SOCK_DGRAM,
+    SOCK_STREAM, SYSPROTO_CONTROL, UTUN_OPT_IFNAME,
+};
+use libs_common::{Error, Result};
+use std::{
+    ffi::{c_int, c_short, c_uchar},
+    io,
+    mem::{size_of, size_of_val},
+    os::fd::{AsRawFd, RawFd},
+    sync::Arc,
+};
+
+use super::InterfaceConfig;
+
+const CTRL_NAME: &[u8] = b"com.apple.net.utun_control";
+const SIOCGIFMTU: u64 = 0x0000_0000_c020_6933;
+
+#[derive(Debug)]
+pub(crate) struct IfaceConfig(pub(crate) Arc<IfaceDevice>);
+
+#[derive(Debug)]
+pub(crate) struct IfaceDevice {
+    fd: RawFd,
+}
+
+impl AsRawFd for IfaceDevice {
+    fn as_raw_fd(&self) -> RawFd {
+        self.fd
+    }
+}
+
+impl Drop for IfaceDevice {
+    fn drop(&mut self) {
+        unsafe { close(self.fd) };
+    }
+}
+// For some reason this is not available in libc for darwin :c
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub struct ifreq {
+    ifr_name: [c_uchar; IF_NAMESIZE],
+    ifr_ifru: IfrIfru,
+}
+
+#[repr(C)]
+union IfrIfru {
+    ifru_addr: sockaddr,
+    ifru_addr_v4: sockaddr_in,
+    ifru_addr_v6: sockaddr_in,
+    ifru_dstaddr: sockaddr,
+    ifru_broadaddr: sockaddr,
+    ifru_flags: c_short,
+    ifru_metric: c_int,
+    ifru_mtu: c_int,
+    ifru_phys: c_int,
+    ifru_media: c_int,
+    ifru_intval: c_int,
+    ifru_wake_flags: u32,
+    ifru_route_refcnt: u32,
+    ifru_cap: [c_int; 2],
+    ifru_functional_type: u32,
+}
+
+// On Darwin tunnel can only be named utunXXX
+pub fn parse_utun_name(name: &str) -> Result<u32> {
+    if !name.starts_with("utun") {
+        return Err(Error::InvalidTunnelName);
+    }
+
+    match name.get(4..) {
+        None | Some("") => {
+            // The name is simply "utun"
+            Ok(0)
+        }
+        Some(idx) => {
+            // Everything past utun should represent an integer index
+            idx.parse::<u32>()
+                .map_err(|_| Error::InvalidTunnelName)
+                .map(|x| x + 1)
+        }
+    }
+}
+
+impl IfaceDevice {
+    fn write(&self, src: &[u8], af: u8) -> usize {
+        let mut hdr = [0, 0, 0, af];
+        let mut iov = [
+            iovec {
+                iov_base: hdr.as_mut_ptr() as _,
+                iov_len: hdr.len(),
+            },
+            iovec {
+                iov_base: src.as_ptr() as _,
+                iov_len: src.len(),
+            },
+        ];
+
+        let msg_hdr = msghdr {
+            msg_name: std::ptr::null_mut(),
+            msg_namelen: 0,
+            msg_iov: &mut iov[0],
+            msg_iovlen: iov.len() as _,
+            msg_control: std::ptr::null_mut(),
+            msg_controllen: 0,
+            msg_flags: 0,
+        };
+
+        match unsafe { sendmsg(self.fd, &msg_hdr, 0) } {
+            -1 => 0,
+            n => n as usize,
+        }
+    }
+
+    pub fn new(name: &str) -> Result<Self> {
```

Comment:
I don't believe we can create sockets on iOS / macOS due to Network Extension limitations. I think we'll have to employ the same approach used by WireGuardKit here: https://git.zx2c4.com/wireguard-apple/tree/Sources/WireGuardKit/WireGuardAdapter.swift#n60